src/netlink.c

```diff
@@ -383,10 +405,190 @@ static int global_create(lua_State *L) {
+    int err;
+    struct nlattr *nl_na;
+    int nl_family_id = -1;
+
+    sprintf(family_name, "NFLUA");
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{unconnected}", 1);
+    p_timeout tm = &nl->tm;
+
+    memset(nl->nlgb, 0, sizeof(struct nlgenmsgbuf));
+
+    nl->nlgb->n.nlmsg_type = GENL_ID_CTRL;
+    nl->nlgb->n.nlmsg_flags = NLM_F_REQUEST;
+    nl->nlgb->n.nlmsg_seq = 0;
+    nl->nlgb->n.nlmsg_pid = luaL_checkinteger(L, 3);
+    nl->nlgb->n.nlmsg_len = NLMSG_LENGTH(GENL_HDRLEN);
+    nl->nlgb->g.cmd = CTRL_CMD_GETFAMILY;
+    nl->nlgb->g.version = 0x1;
+
+    nl_na = (struct nlattr *) GENLMSG_DATA(nl->nlgb);
+    nl_na->nla_type = CTRL_ATTR_FAMILY_NAME;
+    nl_na->nla_len = strlen(family_name) + 1 + NLA_HDRLEN;
+    strcpy(NLA_DATA(nl_na), family_name);
+    nl->nlgb->n.nlmsg_len += NLMSG_ALIGN(nl_na->nla_len);
+
+    memset(&nl_address, 0, sizeof(nl_address));
+    nl_address.nl_family = AF_NETLINK;
+    nl_address.nl_pid = luaL_checkinteger(L, 3);
+
+    socklen_t len = sizeof(nl_address);
+    timeout_markstart(tm);
+    err = socket_sendto(&nl->fd, (char *) nl->nlgb, nl->nlgb->n.nlmsg_len, &sent,
+            (SA *)&nl_address, len, tm);
+    if (err != IO_DONE) {
+        goto out;
+    }
+
+    memset(nl->nlgb, 0, sizeof(struct nlgenmsgbuf));
+    timeout_markstart(tm);
+    err = socket_recvfrom(&nl->fd, (char *)nl->nlgb, NLMSG_SPACE(MAX_PAYLOAD) + GENL_HDRLEN, &sent,
+            (SA *)&nl_address, &len, tm);
+    if (err != IO_DONE && err != IO_CLOSED) {
+        goto out;
+    }
+
+    if (!NLMSG_OK(&nl->nlgb->n, sent)) {
+        goto out;
+    }
+
+    if (nl->nlgb->n.nlmsg_type == NLMSG_ERROR) {
+        goto out;
+    }
+
+    nl_na = (struct nlattr *) GENLMSG_DATA(nl->nlgb);
+    nl_na = (struct nlattr *) ((char *) nl_na + NLA_ALIGN(nl_na->nla_len));
+    if (nl_na->nla_type == CTRL_ATTR_FAMILY_ID) {
+        nl_family_id = *(__u16 *) NLA_DATA(nl_na);
+    }
+
+out:
+    return nl_family_id;
+}
+
+static int meth_receivefrom_generic_nflua(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{unconnected}", 1);
+    size_t got;
+    size_t payload_size;
+    int err = 0;
+    struct sockaddr_nl dst;
+    p_timeout tm = &nl->tm;
+
+    memset(&dst, 0, sizeof(struct sockaddr_nl));
+    memset(nl->nlgb, 0, sizeof(struct nlgenmsgbuf));
+
+    socklen_t len = sizeof(struct sockaddr_nl);
+    timeout_markstart(tm);
+    err = socket_recvfrom(&nl->fd, (char *)nl->nlgb, NLMSG_SPACE(MAX_PAYLOAD) + GENL_HDRLEN, &got,
+            (SA *)&dst, &len, tm);
+    if (err != IO_DONE && err != IO_CLOSED) {
+        lua_pushnil(L);
+        lua_pushstring(L, socket_strerror(err));
+        return 2;
+    }
+
+    if (!NLMSG_OK(&nl->nlgb->n, got)) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "invalid message length");
+        return 2;
+    }
+
+    if (nl->nlgb->n.nlmsg_type == NLMSG_ERROR) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "received message error");
```

Comment:
Is it possible to get an error code or something out of the message?

Reply:
we can get err return value from socket_recvfrom(), the real errno from recvfrom() is also hidden in in some cases . Also few lines before we do print the err ` if (err != IO_DONE && err != IO_CLOSED)`

Reply:
But I guess `recvfrom` and co. are successful here since we did receive a message, it's just that the message itself contains an error? I was wondering if the message of type `NLMSG_ERROR` contains any further info on what the error actually was.

Reply:
Okay kernel has this comment:  `
        /* Error messages get the original request appened, unless the user
         * requests to cap the error message, and get extra error data if
         * requested.
         */
`

I'm not sure what is that when message is originated from nflua. But anyway payload has some string...well we can lua_pushstring() that string. I'll add it.